### PR TITLE
[AutoDiff upstream] Add `SILFunctionType` utilities.

### PR DESCRIPTION
**Pull request by dan-zheng**
Upstream AutoDiff-related `SILFunctionType` utilities:
- `SILFunctionType::getDifferentiabilityParameterIndices`
- `SILFunctionType::getWithDifferentiability`
- `SILFunctionType::getWithoutDifferentiability`

Resolves TF-1125.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -4431,6 +4439,22 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
 
   const clang::FunctionType *getClangFunctionType() const;
 
+  /// Given that `this` is a `@differentiable` or `@differentiable(linear)`
+  /// function type, returns an `IndexSubset` corresponding to the
+  /// differentiability/linearity parameters (e.g. all parameters except the
+  //// `@noDerivative` ones).
```
Comment (marcrasi) [inline include/swift/AST/Types.h]: extra "/"

**Review by marcrasi — APPROVED** [review]

File: include/swift/AST/Types.h (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/Types.h]: Thanks! Fixed in https://github.com/apple/swift/pull/29674/commits/d0b77eb98c3d8f0bbd15bc6901b906bcace4cb79.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Review by rxwei — APPROVED** [review]

**Pull request merged**